### PR TITLE
support Junit5 parallel execution #1472

### PR DESCRIPTION
**Pull request by fladdimir**
resolves https://github.com/microsoft/vscode-java-test/issues/1472

### CODE REVIEW

File: src/runners/junitRunner/JUnitRunnerResultAnalyzer.ts
Comment (fladdimir) [inline src/runners/junitRunner/JUnitRunnerResultAnalyzer.ts:68]: (not all changes to this method are strictly necessary, but I felt it becoming a bit lengthy now, so I extracted some details into separate methods)

**Comment by fladdimir** [comment]
(https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/378 will need to be resolved before this feature will work correctly)

**Comment by jdneo** [comment]
@fladdimir Thank you for taking look into this and deal with the upstream issue.

Will it be fine if I review this PR after the upstream changes are merged?

**Comment by fladdimir** [comment]
> Will it be fine if I review this PR after the upstream changes are merged?

@jdneo of course thats fine, thanks for the quick reply

**Comment by fladdimir** [comment]
Manual tests (including the locally built upstream fix for the RemoteTestRunner):

before - test2 appears not to be run at all, swallowing its failure:
[before.webm](https://user-images.githubusercontent.com/53265832/211956161-56b5a43b-ccf6-4455-a0fc-263de5e83243.webm)

fixed - both test failures are shown, comparison failures and traces are placed correctly:
[fixed.webm](https://user-images.githubusercontent.com/53265832/211956166-6e21d0ce-cbb3-43be-bafc-a5514edc05af.webm)

**Comment by fladdimir** [comment]
@jdneo the [upstream PR](https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/379) has been merged.

Unfortunately I am not familiar with the workflow. I assume the upstream artifact (`org.eclipse.jdt.junit.runtime_3.7.x.jar`) needs to be published somewhere to become part of the plugins of a `eclipse.jdt.ls` snapshot build, which could then be included into an upcoming `vscode-java` extension (preview) release?
(In case that is the way it works, do you have an idea of what could be a timeline for that?)

**Comment by jdneo** [comment]
Cool! I saw the upstream PR has been merged, great job!

The workflow looks like in this way:

1. The upstream JDT will first release an I-build which includes your changes. I noticed the [latest I-build](https://download.eclipse.org/eclipse/updates/4.27-I-builds/I20230112-1800/plugins/) should have already included your changes.
2. Then I will raise a PR in this repo to update the dependencies and pick up the latest bits.
3. Rebase and Review your PR, merge it.
4. After your PR is merged, it will be available very soon in the extension pre-release channel.

**Comment by jdneo** [comment]
Cool! I saw the upstream PR has been merged, great job!

The workflow looks like in this way:

1. The upstream JDT will first release an I-build which includes your changes. I noticed the [latest I-build](https://download.eclipse.org/eclipse/updates/4.27-I-builds/I20230112-1800/plugins/) should have already included your changes.
2. Wait for the vscode-java update the dependency to include the `org.eclipse.jdt.junit.runtime` which contains your change.
3. Review your PR, merge it.
4. After your PR is merged, it will be available very soon in the extension pre-release channel.

I can help update the `vscode-java` side next week

File: test/test-projects/junit/src/test/java/junit5/ParallelExecutionTest.java
Comment (jdneo) [inline test/test-projects/junit/src/test/java/junit5/ParallelExecutionTest.java:26]: Is the code here be commented out on purpose?

**Comment by fladdimir** [comment]
> 1. The upstream JDT will first release an I-build which includes your changes. I noticed the [latest I-build](https://download.eclipse.org/eclipse/updates/4.27-I-builds/I20230112-1800/plugins/) should have already included your changes.
> 2. Wait for the vscode-java update the dependency to include the `org.eclipse.jdt.junit.runtime` which contains your change.
> I can help update the `vscode-java` side next week

That sounds great, thanks a lot!

File: test/test-projects/junit/src/test/java/junit5/ParallelExecutionTest.java
Comment (fladdimir) [reply test/test-projects/junit/src/test/java/junit5/ParallelExecutionTest.java]: not really - I removed the commented-out code and updated the class so that it now includes two failing tests as well as a passing one

**Comment by Kropie** [comment]
Awesome work @fladdimir! When I was looking into this I was noticing the issues regarding the interleaving of messages from the eclipse provided unit test runner - great job finding the cause of the issue and working to resolve it!

**Comment by jdneo** [comment]
I raised a PR to update the dependency in VS Code Java: https://github.com/eclipse/eclipse.jdt.ls/pull/2403.

since it's about to release a new version this week, that PR should be able to be merged after the release (late this week)

File: src/runners/junitRunner/JUnitRunnerResultAnalyzer.ts
Comment (jdneo) [inline src/runners/junitRunner/JUnitRunnerResultAnalyzer.ts:24]: nit: how about give it a more specific name, something like: `tracingItem`?

**Review by jdneo — COMMENTED** [review]
I looked the PR today and overall it looks good! 👍

After the latest pre-release of vscode-java is available, I'll play with it manually and merge this PR if everything works fine.

File: src/runners/junitRunner/JUnitRunnerResultAnalyzer.ts
Comment (fladdimir) [reply src/runners/junitRunner/JUnitRunnerResultAnalyzer.ts]: just changed the name 👍🏼

**Comment by jdneo** [comment]
I found that there will be two assertion failure message when running the added sample tests

image

Can we deduplicate then to single one?

**Comment by fladdimir** [comment]
> I found that there will be two assertion failure message when running the added sample tests
> Can we deduplicate then to single one?

While debugging I realized that this duplication of diff-messages also occurs during normal test execution on the current main branch, without being related to parallel test execution.

Apparently its a problem of the general stack trace processing, occuring when a failing equality assertion stack trace contains multiple entries from the test class(?).

(Non-concurrently executed) sample based on the current main branch:

test_duplicated_diff

https://github.com/microsoft/vscode-java-test/pull/1521

There is also a unit-test (part of `JUnitAnalyzer.test.ts`) to further debug this issue, which reveals that the following is called with a new diff-message each time a matching stack-trace entry is processed:
https://github.com/microsoft/vscode-java-test/blob/a3848d4145056d72d9ac50b4cc2efe4a90df2593/src/runners/baseRunner/RunnerResultAnalyzer.ts#L60 seems to be 

I'd like to continue investigating, however, would it be ok to open a separate issue for this?

**Comment by jdneo** [comment]
Sorry for the late reply. I'm on vacation these days :)

Yes, I can confirm that this bug can also be observed in the marketplace version. And it's totally fine to deal with it in a separate issue.

Thank you for the contribution!

**Review by jdneo — APPROVED** [review]

**Comment by fladdimir** [comment]
> Sorry for the late reply. I'm on vacation these days :)
> 
> Yes, I can confirm that this bug can also be observed in the marketplace version. And it's totally fine to deal with it in a separate issue.
> 
> Thank you for the contribution!

no worries at all - thanks for the review and merge!
enjoy your holidays!

**Pull request merged**
